rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,350 @@
+ * Delete old `"activity"` sequence entry
+ * Update `JobOpts`
+ * Create a new `"activity"` sequence entry and in main job table
+ * Update `"watches"` sequence for that job type
+
+`finish()`:
+ * Check if `WorkerLockId` matches, otherwise returns `worker_conflict`
+ * Check if `Canceled` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
+ * If `"resubmit"` field in `JobOpts` is `true`, re-enqueue the job into the pending queue.
+ * Set job table's `WorkerLockId` to `null`
+
+## Activity Monitor Implementation
+
+There is an activity monitor running for each job type. It performs periodic
+scans of the `"activity"` range with sequences less than `"watches"` sequence for
+that type:
+
+  * On first run:
+    * Read the `ActivityTimeout` value and remember the `"watches"` sequence
+  * On subsequent runs:
+    * Call `get_range()` on `"activity"` entries less than the remembered sequence
+    * Read and store the new `"watches"` sequence
+    * If there are any stale entries returned, re-enqueue them
+
+Every time a job is updating its `"activity"`, it bumps the `"watches"`
+sequence. That means `"watches"` sequence points to the newest updated sequence
+for that job type at that point in time. That allows anyone, in this case the
+activity monitor, to remember the `"watches"` sequence value then return some
+time later, and query if anything has not been updated since time it saw that
+sequence. If everything is working properly and jobs are updating that should
+return an empty list.
+
+## Job State Diagram
+
+The job state transition diagram looks something like this:
+
+```
+             +--------- accept() ---------->+
+             |                              |
+             |                              v
+add() -> [PENDING]                      [RUNNING] ---> finish(), resubmit=false --> [FINISHED]
+             ^                              |
+             |                              |
+             +-- finish(), resubmit=true --+
+             ^                              |
+             |                              |
+             +-- activity monitor timeout --+
+```
+
+When a job is added it immediately becomes `pending`. When it is `accepted()`
+for execution by a worker it becomes `running`. Then, if `finish` is called it
+becomes finished if the `"resubmit"` option is `false`, or it goes back to
+`pending` if it is `true`. A job may also be moved from `running` to
+`pending` by the activity monitor if the job fails to update its status often
+enough.
+
+Each job type MAY define additional states. For example, replication jobs MAY
+have both `completed` and `failed` as terminal states. In both cases the
+framework will consider those as `finished`, but job-type specific data MAY
+indicate that `completed` is finished with a successful result, and `failed` is
+finished with a terminal failure (say an invalid replication doc format).
```

Comment:
> This discussion on job states is confusing to me. Referencing the data model I don't see a state variable anywhere except for the `pending` subspace. Which has me wondering where the running/complete/failed/finished etc states come into effect. I could see running as implicit by its inclusion in the `activity` subspaces. Though that also seems like it might be better off as `active` as the state name.

The states are implicit depending if they are pending or running. Running jobs have a worker lock id. I'll add the following to the `Job State Diagram` 

> These states are not tracked via an explicit state label. If the job is in the
pending queue, it is considered pending, if it has a lock id acquired by a job
processor, it is considered running.

How does that sound?

> I'm also not clear on what happens to a job after its "done" (abstractly, i.e., all of completed/failed/finished states). You've mentioned replications here as a reference which is a one approach since we maintain info about them so that users will realize why a replication isn't finished. But a secondary index update at some point needs to disappear from the database at some point right? Who makes that decision when? Or perhaps the "terminal" state here is something that's been pulled too far in from the replicator and maybe shouldn't be part of this RFC?

I'll remove the discussion about the terminal states. A job that is finished is not auto-removed, it stays in the jobs table. The job creator can inspect the result and remove it.  The worker maybe also remove the job when finished if the job wasn't re-enqueued.